# Import libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Importing data

In [3]:
path = r'C:\Users\justi\OneDrive\Documents\Career foundry\Data Immersion 4'

In [5]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Data consistency prods

In [8]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [9]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [10]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [11]:
df_prods.shape

(49693, 5)

In [13]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [14]:
df_prods_clean.shape

(49677, 5)

In [15]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [16]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

In [21]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# Data consistency ords

## Check values

In [22]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_num_of_customer  \
count  3.421083e+06  3.421083e+06  3.421083e+06           3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05           1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04           1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00           1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04           5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05           1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05           2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05           1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

In [23]:
df_ords

Unnamed: 0  order_id  user_id  order_num_of_customer  \
0                 0   2539329        1                      1   
1                 1   2398795        1                      2   
2                 2    473747        1                      3   
3                 3   2254736        1                      4   
4                 4    431534        1                      5   
...             ...       ...      ...                    ...   
3421078     3421078   2266710   206209                     10   
3421079     3421079   1854736   206209                     11   
3421080     3421080    626363   206209                     12   
3421081     3421081   2977660   206209                     13   
3421082     3421082    272231   206209                     14   

         orders_day_of_week  order_hour_of_day  days_since_last_order  
0                         2                  8                    NaN  
1                         3                  7                   15.0  
2                         3                 12                   21.0  
3                         4                  7                   29.0  
4                         4                 15                   28.0  
...                     ...                ...                    ...  
3421078                   5                 18                   29.0  
3421079                   4                 10                   30.0  
3421080                   1                 12                   18.0  
3421081                   1                 12                    7.0  
3421082                   6                 14                   30.0  

[3421083 rows x 7 columns]

In [33]:
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [34]:
# the only thing I would look at more is a customer having 100 orders. The unnnamed: 0 column is not needed

## Check for any mixed-type data

In [35]:
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_ords[weird]) > 0:
        print(col)

In [29]:
# Check for missing values

## Check for missing values

In [36]:
df_ords.isnull().sum()

order_id                      0
user_id                       0
order_num_of_customer         0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [37]:
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [38]:
df_ords_nan

order_id  user_id  order_num_of_customer  orders_day_of_week  \
0         2539329        1                      1                   2   
11        2168274        2                      1                   2   
26        1374495        3                      1                   1   
39        3343014        4                      1                   6   
45        2717275        5                      1                   3   
...           ...      ...                    ...                 ...   
3420930    969311   206205                      1                   4   
3420934   3189322   206206                      1                   3   
3421002   2166133   206207                      1                   6   
3421019   2227043   206208                      1                   1   
3421069   3154581   206209                      1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 6 columns]

These missing values are becaue there were no days since the last order. Meaning they are a new customer

In [40]:
df_ords[df_ords['order_num_of_customer'] ==1]

order_id  user_id  order_num_of_customer  orders_day_of_week  \
0         2539329        1                      1                   2   
11        2168274        2                      1                   2   
26        1374495        3                      1                   1   
39        3343014        4                      1                   6   
45        2717275        5                      1                   3   
...           ...      ...                    ...                 ...   
3420930    969311   206205                      1                   4   
3420934   3189322   206206                      1                   3   
3421002   2166133   206207                      1                   6   
3421019   2227043   206208                      1                   1   
3421069   3154581   206209                      1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 6 columns]

This table showing the order_num_of_customer = 1 proves all these instances (206209 rows) are when a user places an order for the first time.
The first table showing all of the NaN is equal to all the first time orders. 

In this case I think the best thing to do is to leave all of these as missing. That way I can still tell when someone places their first order.
Along with this I can still tell if someone places multiple orders in a day, which would have a 0 in the 'days_since_last_order column. 

## Check for duplicate values

In [41]:
df_dups = df_ords[df_ords.duplicated()]

In [42]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, order_num_of_customer, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

There are no duplicate values

## Export

In [43]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))